### PR TITLE
Feature/convert old components to fetch state

### LINKED ISSUE #636: Convert old component to fetch state
## Overview
Convert old component using a `'loading' | 'ready' | 'error'` state + some data to `FetchState`

## Acceptance criteria
Convert:
- [ ] GroupManagerListComponent
- [ ] AddItemComponent (just remove the state?)
- [ ] ChapterChildrenComponent : add the validation thing in the data, so map the data to `{ children: ... , missingValidation: bool }`
- [ ] ItemChildrenEditComponent
- [ ] ParentSkillsComponent
- [ ] SubSkillsComponent
- [ ]

### LINKED ISSUE #734: Bug: Error on visiting manager page
When visiting manager pages (for instance https://dev.algorea.org/en/#/groups/by-id/672913018859223173;path=52767158366271444/details/managers), I get a ` ERROR TypeError: t.createEmbeddedView is not a function` in the console.

### PR DESCRIPTION
**Pull request by Iloveall**
Fixes #636
Fixes #734

### CODE REVIEW

File: src/app/modules/group/components/group-manager-list/group-manager-list.component.ts [hunk 1]
```diff
@@ -1,28 +1,39 @@
...
-
   selection: Manager[] = [];
   removalInProgress = false;
   isPermissionsEditDialogOpened = false;
   dialogManager?: Manager & { canManageAsText: string };
 
+  private refresh$ = new Subject<void>();
+  private readonly groupId$ = new ReplaySubject<string>(1);
+  readonly state$ = this.groupId$.pipe(
+    distinct(),
```
Comment (ThomasJuster) [inline src/app/modules/group/components/group-manager-list/group-manager-list.component.ts:27]: Although the case will probably not happen, we should use `distinctUntilChanged` here:
If the id is `1`, then `2` and then `1` again, we want to refetch item 1

File: src/app/modules/group/components/group-manager-list/group-manager-list.component.ts [hunk 2]
```diff
@@ -31,7 +42,13 @@ export class GroupManagerListComponent implements OnChanges {
...
 
   ngOnChanges(_changes: SimpleChanges): void {
-    this.reloadData();
+    if (this.groupData) {
+      this.groupId$.next(this.groupData.group.id);
+    }
+  }
+
+  ngOnDestroy(): void {
+    this.refresh$.complete();
```
Comment (ThomasJuster) [inline src/app/modules/group/components/group-manager-list/group-manager-list.component.ts:51]: `groupId$` should be completed as well

File: src/app/modules/item/components/chapter-children/chapter-children.component.html [hunk 3]
```diff
@@ -1,18 +1,19 @@
...
-    To validate this chapter, solve at least all tasks with Validation type
-  </p>
-  <p class="validation-text" *ngIf="state === 'empty'" i18n>
-    This chapter has no content visible to you, so you can't validate it for now.
-  </p>
+  <ng-container *ngIf="state.isReady && state.data as data">
+    <br>
+    <p class="validation-text" *ngIf="data.missingValidation" i18n>
+      To validate this chapter, solve at least all tasks with Validation type
+    </p>
```
Comment (ThomasJuster) [inline src/app/modules/item/components/chapter-children/chapter-children.component.html:10]: I did not find a test case for this paragraph, do you have one ? :pray:

File: src/app/modules/item/components/chapter-children/chapter-children.component.ts [hunk 4]
```diff
@@ -24,18 +26,47 @@ interface ItemChildAdditions {
...
   @Input() itemData?: ItemData;
 
-  state: 'loading' | 'error' | 'empty' | 'ready' | 'ready-missing-validation' = 'loading';
-  children: (ItemChild&ItemChildAdditions)[] = [];
+  private readonly params$ = new ReplaySubject<{ id: string, attemptId: string }>(1);
+  readonly state$ = this.params$.pipe(
+    distinctUntilChanged((a, b) => a.id === b.id && a.attemptId === b.attemptId),
+    switchMap(({ id, attemptId }) =>
+      this.getItemChildrenService.get(id, attemptId).pipe(
+        map((itemChildren: ItemChild[]) => {
```
Comment (ThomasJuster) [inline src/app/modules/item/components/chapter-children/chapter-children.component.ts:34]: Unless I'm missing something, there's no need to nest the `map` operator into the `switchMap`
When possible, we should keep observable declarations as flat as possible for readability

Plus, considering the size of the `map`, I think some of its content can be externalised into class methods

File: src/app/modules/item/components/item-children-edit/item-children-edit.component.ts [hunk 5]
```diff
@@ -118,22 +111,44 @@ export class ItemChildrenEditComponent implements OnChanges {
...
+      if (state.isReady) {
+        this.data = state.data;
+        this.scoreWeightEnabled = this.data.some(c => c.scoreWeight !== 1);
+        this.onChildrenListUpdate();
+      }
+    });
+  }
+
+  ngOnDestroy(): void {
+    this.refresh$.complete();
```
Comment (ThomasJuster) [inline src/app/modules/item/components/item-children-edit/item-children-edit.component.ts:135]: `params$` should be completed as well

File: src/app/modules/item/components/item-children-edit/item-children-edit.component.ts [hunk 6]
```diff
@@ -66,21 +67,13 @@ export class ItemChildrenEditComponent implements OnChanges {
...
-        .getWithInvisibleItems(this.itemData.item.id, this.itemData.currentResult.attemptId)
+  private readonly params$ = new ReplaySubject<{ id: string, attemptId: string }>(1);
+  private readonly refresh$ = new Subject<void>();
+  readonly state$: Observable<FetchState<PossiblyInvisibleChildData[]>> = this.params$.pipe(
+    distinctUntilChanged((a, b) => a.id === b.id && a.attemptId === b.attemptId),
+    switchMap(({ id, attemptId }) =>
+      this.getItemChildrenService
+        .getWithInvisibleItems(id, attemptId)
         .pipe(
           map(children => children
```
Comment (ThomasJuster) [inline src/app/modules/item/components/item-children-edit/item-children-edit.component.ts:78]: Same remark as upper, the `map` can be declared after `switchMap` and not inside
(Unless I'm missing something)

File: src/app/modules/item/components/parent-skills/parent-skills.component.ts [hunk 7]
```diff
@@ -19,18 +20,29 @@ interface ParentSkillAdditions {
...
-  state: 'loading' | 'error' | 'ready' = 'loading';
-  parents: (ItemParent&ParentSkillAdditions)[] = [];
-
-  private subscription?: Subscription;
+  private readonly params$ = new ReplaySubject<{ id: string, attemptId: string }>(1);
+  readonly state$ = this.params$.pipe(
+    distinctUntilChanged((a, b) => a.id === b.id && a.attemptId === b.attemptId),
+    switchMap(({ id, attemptId }) =>
+      this.getItemParentsService.get(id, attemptId).pipe(
+        map(parents => parents.map(parent => ({ ...parent, isLocked: !canCurrentUserViewItemContent(parent) })))
```
Comment (ThomasJuster) [inline src/app/modules/item/components/parent-skills/parent-skills.component.ts:28]: Same remark about the nest `map` operator

File: src/app/modules/item/components/sub-skills/sub-skills.component.ts [hunk 8]
```diff
@@ -24,18 +25,43 @@ interface SubSkillAdditions {
...
-  state: 'loading' | 'error' | 'ready' = 'loading';
-  children: (ItemChild&SubSkillAdditions)[] = [];
-
-  private subscription?: Subscription;
+  private readonly params$ = new ReplaySubject<{ id: string, attemptId: string }>(1);
+  readonly state$ = this.params$.pipe(
+    distinctUntilChanged((a, b) => a.id === b.id && a.attemptId === b.attemptId),
+    switchMap(({ id, attemptId }) =>
+      this.getItemChildrenService.get(id, attemptId).pipe(
+        map(children =>
```
Comment (ThomasJuster) [inline src/app/modules/item/components/sub-skills/sub-skills.component.ts:33]: same old same old :smile: 
I point them out to make sure we don't forget one, that's all :grimacing:

**Review by ThomasJuster — COMMENTED** [review]
Nice, everything works as expected :+1: 

Most of the remarks are actually the same and only minor ones

File: src/app/modules/group/components/group-manager-list/group-manager-list.component.ts (hunk 1 above)
Comment (smadbe) [reply src/app/modules/group/components/group-manager-list/group-manager-list.component.ts:27]: Oops, I mixed both as well. I think most of usage of `distinct` in the project should be `distinctUntilChanged`. 
I create an issue...

Comment (Iloveall) [reply src/app/modules/group/components/group-manager-list/group-manager-list.component.ts]: Fixed

File: src/app/modules/item/components/sub-skills/sub-skills.component.ts (hunk 8 above)
Comment (Iloveall) [reply src/app/modules/item/components/sub-skills/sub-skills.component.ts]: Fixed

File: src/app/modules/item/components/item-children-edit/item-children-edit.component.ts [hunk 9]
```diff
@@ -118,22 +111,44 @@ export class ItemChildrenEditComponent implements OnChanges {
...
+    mapToFetchState({ resetter: this.refresh$ }),
+  );
+
+  constructor(
+    private getItemChildrenService: GetItemChildrenService,
+    private itemRouter: ItemRouter,
+  ) {}
+
+  ngOnInit(): void {
+    this.subscription = this.state$.subscribe(state => {
```
[the remark this answers is not in the document]
Comment (ThomasJuster) [reply src/app/modules/item/components/item-children-edit/item-children-edit.component.ts:120]: I meant inline, like `this.state$.pipe(readyData()).subscribe(…)`
But anyway, it's not important, just resolve after reading :+1:

**Review by ThomasJuster — APPROVED** [review]

File: src/app/modules/item/components/item-children-edit/item-children-edit.component.ts (hunk 9 above)
[the remark this answers is not in the document]
Comment (Iloveall) [reply src/app/modules/item/components/item-children-edit/item-children-edit.component.ts:120]: I see, Okay :)

**Pull request merged**